### PR TITLE
Fill patches in add_patch and flatten their real geometry

### PR DESCRIPTION
**Pull request by michael-denyer**
`Axes.add_patch` draws every patch as edge-only line segments. Its docstring admits the outline approximation, but the gap is wider than that, and I found two further problems while fixing the fill.

Measured against Matplotlib 3.11.1:

| Patch | xy before | Correct |
|---|---|---|
| `Rectangle(facecolor="tab:blue")` | hollow | filled |
| `Rectangle((1,2), 3, 4, angle=30)` | x spans 1.000..4.000, `angle` dropped | x spans -1.000..3.598 |
| `Circle(radius=1)` | area 3.2509 | 3.1416 |
| `Ellipse(width=2, height=1, angle=20)` | area 3.2509 | 1.5708 |

The rotation and curve errors have the same root cause as the missing fill. `add_patch` reconstructs geometry by hand from `get_x`/`get_y`/`get_width`/`get_height`, which has nowhere to put `angle`, and falls back to raw `path.vertices`, which are Bezier control points rather than the curve. The ellipse is out by a factor of two because nothing applies the patch transform.

## What changed

Geometry now comes from `patch.get_path().to_polygons(patch.get_patch_transform())`, the same flattening Matplotlib's own renderers use. It applies the transform and flattens curves, so one call replaces the hand-rolled branching and fixes all three symptoms at once.

Each ring then fills through the recipe `Axes.fill` already uses, `kernels.polygon_triangles` into a `triangle_mesh` mark with `_joined_fill=True`. No new mark and no Rust change.

Extraction and paint are now separate. `_patch_outline` returns data-space rings, `_patch_fill_color` returns the face color or `None`, and `add_patch` runs one fill pass and one outline pass over them. Conflating extraction with a single edge-only emit is what let three distinct bugs share one line of code.

A filled patch emits more than one mark, so `add_patch` returns a `Patch` handle that carries the outline entries and sweeps them in `remove()` and `set_zorder()`. This follows `Wedge` in `_artists.py`. Without it the change would regress `remove()`, which used to clear a patch completely and would otherwise leave a ghost outline behind.

Outlines now take the patch's own `get_linewidth()` instead of a fixed 1.0.

## Preserved behaviour

Patches reporting `fill=False`, a `"none"` face color, or a fully transparent one stay edge-only. `StepPatch` still routes to `stairs`. Unsupported objects still raise `TypeError`. The axes color cycle is never advanced, because the fill only ever uses the patch's own face color. A degenerate ring such as a zero-height `Rectangle` has no triangulation, so it draws its edge and skips the fill rather than raising.

## Evidence

`tests/pyplot` goes from 1174 to 1183 passing, 3 skipped, and `make check` passes with 3443 passing. The `ty` advisory count is 77 both before and after.

Nine tests added, five of which fail against `main`:

```
FAILED test_added_rectangle_patch_fills_with_its_own_face_color
FAILED test_added_rotated_rectangle_keeps_its_rotation
FAILED test_added_ellipse_flattens_its_curve_under_the_patch_transform
FAILED test_added_concave_polygon_patch_triangulates_its_true_area
FAILED test_removing_a_filled_patch_takes_its_outline_with_it
```

The other four are regression guards that pass either way. The color-cycle guard is one of them, so a plain before-and-after run proves nothing about it. Swapping `resolve_color(face)` for `self._next_patch_color()` makes it fail with `+ #ff7f0e`, which is the mistake worth guarding given how close `_next_patch_color` sits to this code.

Rendered output, not just entry structure. Exporting a rotated `Rectangle`, a concave `Polygon`, and an `Ellipse` at 100 dpi and counting pixels by color, using the polygon as the scale reference:

```
rot_rect  18808 px, true area 12.0000, predicted 18616, 1.03% error
concave   15513 px, true area 10.0000, predicted 15513, 0.00% error
ellipse    2413 px, true area  1.5708, predicted  2437, 0.98% error
```

Area alone cannot prove rotation, since an unrotated 3x4 rectangle also covers 12. Measuring the rectangle's x extent against the polygon's known 4-unit span gives 4.588 data units. Rotated is 4.598 and unrotated would be 3.000. In SVG the concave shape arrives as a single 5-vertex `<polygon>` rather than three triangles, which confirms `_joined_fill` suppresses internal seams through to the exported file.

## Two related bugs left alone

`Axes.fill` orphans its own edge mark the same way, and filled `stairs` orphans its hatch mark. Both predate this change, and `Patch` now gives them a precedent to copy. Fixing them here would span concerns the contributor guide asks to keep apart.

Separately, `add_patch` ignores `get_label()` entirely, before and after this change, so a `Rectangle(label="region")` used as a legend proxy never appears. That is a feature gap rather than a regression.

## Docs

`spec/matplotlib/compat-changelog.md` gets a dated entry, `CHANGELOG.md` a `Fixed` entry, and `spec/matplotlib/compat.md` one line, since curve flattening through `to_polygons` belongs in its list of accepted visual approximations. `spec/matplotlib/shim-todo.md` needs nothing. Line 396 already reads "wider `add_patch` / `add_collection` mappings for common Matplotlib objects", which stays true.

No benchmarks. This is a correctness fix and makes no performance claim.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved patch rendering: fills are applied only when enabled and appropriate; `fill=False`, `"none"`, or fully transparent faces stay outline-only.
  * Enhanced patch geometry handling (including curved shapes), with more robust vertex deduplication to avoid ring collapse from non-finite values.
  * Patch outlines now respect the patch `linewidth`, scaled for the output DPI.
* **New Features**
  * `Axes.add_patch` now returns a composite patch handle so `remove()`/`set_zorder()` affect the full patch.
* **Compatibility**
  * Matplotlib-like nested-ring “hole” behavior: holes are not filled.
* **Tests**
  * Added compat tests for non-finite rectangle geometry edge cases.

### CODE REVIEW

**Comment by michael-denyer** [comment]
Both findings tested rather than taken on trust. One is right as stated, one is right about the defect and wrong about the example, and chasing the second turned up a third bug neither of us had. All three are fixed in `d7b441d`.

## Stroke units, confirmed

Right. Thickness in pixels of a 5 pt stroke, measured through the centre column of a rendered PNG:

```
                before                     after
line   dpi= 72  [5]      dpi=200 [13]      unchanged
patch  dpi= 72  [5, 5]   dpi=200 [5, 5]    dpi=72 [5, 5]   dpi=200 [13, 13]
```

`ax.plot` scaled, the patch outline did not move at all. Now multiplied by `self._point_scale()` and it tracks `ax.plot` exactly.

One correction to the framing. Released `add_patch` hardcoded `width` to `1.0`, so it ignored DPI *and* dropped `linewidth` entirely. This PR's earlier commit read the right number in the wrong unit. It is now right in both, and the changelog says so.

## Hole topology, right about the defect, wrong about the example

`Wedge(..., width=...)` was not a good case to cite. An annular sector comes back from `to_polygons` as a single ring that traces the outer arc and returns along the inner one, so it already filled correctly:

```
Wedge(0, 90, width=0.4)     1 mesh, area 0.5022, true 0.5027   correct before and after
```

Compound paths are the real case, and there the finding holds:

```
PathPatch square-with-hole  2 meshes, areas 100.0 and 16.0      paints 116 for a true 84
```

Not implementing hole support here. `kernels.polygon_triangles` takes one simple polygon and rejects anything else at the kernel boundary, so a hole needs either a bridged seam in Python, which `_joined_fill` would show through because a bridge is not a shared triangle edge, or a new kernel entry point. That is a maintainer's design call, not a review response.

Abstaining instead. When rings nest, the patch draws its outline and skips the fill, which is exactly the released behaviour, so nothing regresses and no known-wrong pixels get painted. Containment is the test, not ring count, so disjoint rings still fill. The check reuses `kernels.polygon_select`, the even-odd cast already shipped for lasso selection, rather than hand-rolled ray casting.

```
                            before                    after
square with square hole     2 meshes, 100.0 and 16.0  0 meshes, 2 outlines
disjoint squares            fills both                fills both
annular sector              1 mesh, 0.5022            unchanged
```

## Third bug, found while testing the second

A full annulus emitted no fill at all, rather than the solid sector predicted:

```
Wedge(0, 360, width=0.4)    2 outlines, 0 meshes
```

Root cause. Matplotlib repeats a vertex inside its full-circle paths, so 33 vertices reached the triangulator carrying 32 unique points, and `polygon_triangles` rejects duplicates outright. The `except ValueError` around it then discarded the failure in silence.

```
Circle(1)              ring 0: 16 unique -> 14 triangles
Wedge 0-360 width .4   ring 0: 33 pts, 32 unique -> REJECTED
                       ring 1: 33 pts, 32 unique -> REJECTED
```

Rings now drop every consecutive duplicate, not just a trailing one, and both annulus rings triangulate. They still end up outline-only, but now because the nesting check abstains rather than because an exception vanished. The regression test asserts the ring vertex count for that reason, since `meshes == []` alone cannot tell the two causes apart.

## Verification

`tests/pyplot` 1183 to 1189 passing, 3 skipped. `make check` passes, 8 checks, 3449 passing. The `ty` advisory count is 77, identical to base, so the PR still adds none.

Six tests added, four failing against the previous head:

```
FAILED test_patch_outline_width_converts_points_to_pixels
FAILED test_patch_outline_stroke_thickens_with_dpi_in_the_png
FAILED test_patch_with_nested_rings_abstains_from_filling_the_hole
FAILED test_full_annulus_draws_both_outlines_without_raising
```

The two that pass either way guard the abstention against over-reaching, so passing before and after is the signal they exist for.

## Known limitation, now documented

Nested rings never fill. A compound `PathPatch` with a hole and a full-circle `Wedge` with a `width` draw outlines only. Recorded in the compat changelog, the accepted-approximations line in `compat.md`, the docstring, and `CHANGELOG.md`.

## Unrelated, spotted while here

`Axes.fill` passes its edge linewidth to the mark unscaled at `_plot_types.py:1692`, the same points-versus-pixels bug found here. It also orphans its edge mark on `remove()`, as does filled `stairs` with its hatch mark. All predate this PR and are left alone.

File: python/xy/pyplot/_axes.py [hunk 1]
```diff
@@ -5582,24 +5622,47 @@ def add_patch(self, patch: Any) -> Artist:
...
+                entries.append(
+                    self._add(
+                        "@mark",
+                        {
+                            "factory": "triangle_mesh",
+                            "args": (x0, y0, x1, y1, x2, y2),
+                            "kwargs": {"color": resolve_color(face), "_joined_fill": True},
+                        },
+                    )
+                )
```
[the remark this answers is not in the document]
Comment (michael-denyer) [reply python/xy/pyplot/_axes.py:5844]: Kept open deliberately, with the fill-rule part unimplemented and documented.

The `Wedge(..., width=...)` half of this does not reproduce. An annular sector comes back from `to_polygons` as one ring tracing the outer arc and returning along the inner one, so it filled correctly before this change and still does, at area 0.5022 against a true 0.5027. Compound `PathPatch` holes are the real case, and they painted 116 for a true 84.

Rings that nest now abstain: outline drawn, fill skipped. That is the released behaviour, so nothing regresses and no known-wrong pixels get painted. Containment is the test rather than ring count, so disjoint rings still fill both.

Honouring the fill rule properly needs more than this PR should carry. `kernels.polygon_triangles` accepts one simple polygon and rejects anything else at the kernel boundary, so a hole means either bridging a seam in Python, which `_joined_fill` would show through because a bridge is not a shared triangle edge, or a new kernel entry point. Either is a design decision for a maintainer.

Testing this finding did surface a third bug, fixed in `d7b441d`. A full annulus emitted no fill at all rather than the solid sector predicted here. Matplotlib repeats a vertex inside its full-circle paths, so 33 vertices reached the triangulator carrying 32 unique points, `polygon_triangles` rejected duplicates, and an `except ValueError` swallowed it. Rings now drop every consecutive duplicate.

**Comment by michael-denyer** [comment]
Both fixed in `9e6490e` and `e79b1da`. The first finding was right and the damage is worse than the example suggested, but the suggested remedy would have silently undone an earlier fix, so the implementation differs.

## Tolerance-based deletion, confirmed

Worse than reported. `np.isclose` scales to coordinate magnitude, so at x = 1e9 it treats anything within 10,000 data units as one vertex. Through the real code path, not a simulation:

```
Rectangle(x=1e9, width=5000, height=10)
  before   outline verts [3]   mesh area     0.0
  after    outline verts [5]   mesh area 50000.0
```

The patch did not degrade, it disappeared. Genomic coordinates, epoch milliseconds and financial ticks all sit in that range.

## Why the suggested fix is not the one applied

Deduplicating only exactly identical consecutive vertices catches nothing here. Matplotlib's repeat is floating-point noise, not a coincidence:

```
annulus ring0: exact_dupes=0  isclose_dupes=1
annulus ring1: exact_dupes=0  isclose_dupes=1

with exact-only dedupe:
  ring 0: 33 pts -> REJECTED (polygon must be finite, simple, and non-degenerate)
  ring 1: 33 pts -> REJECTED
```

That is the full-annulus bug from the previous round coming straight back, and it would come back silently.

## What was applied

An absolute tolerance of `1e-12` times the ring's own bounding-box diagonal. Scaling to the ring rather than to coordinate magnitude satisfies both cases, and the threshold is not a judgement call. Gap between consecutive vertices as a fraction of ring span:

```
                    smallest      next smallest
annulus ring0       8.660e-17       6.957e-02
annulus ring1       8.660e-17       6.957e-02
circle              1.400e-01       1.400e-01
rect at x=1e9       2.000e-03       2.000e-03
```

Every repeat is noise at ~1e-16 of the span. The smallest real edge across every shape we flatten is 2e-3. Fourteen orders of magnitude apart, and the threshold sits in the middle. Both populations are recorded in the docstring so the constant stays re-derivable.

Multiplying by span rather than dividing the gaps means a zero-span ring needs no special case; the tolerance becomes zero and the test degrades to exact equality. The comparison is negated so a non-finite gap keeps its vertex and the ring is rejected by the triangulator on its own terms rather than being silently emptied.

After:

```
rect at x=1e9    rings=[5]       area 50000.0000   true 50000.0
full disc Wedge  rings=[33]      area     3.1414   true     3.1416
annulus          rings=[33, 33]  abstains, nested rings
tiny 1e-9 edge   rings=[6]       area     1.0000   true     1.0000
```

## The pinned tessellation count

Right, and removed. One correction worth recording. The obvious replacement, asserting no zero-length segments, is a dead assertion: it passes against the pre-dedupe commit, because a matplotlib repeat is 2.4e-16 apart rather than coincident, so the stub segment has nonzero length. The test now asserts the shortest segment as a fraction of the outline's bounding diagonal exceeds 1e-6, which is scale-free, pins no number we do not control, and fails against the pre-dedupe state as intended.

## Verification

`tests/pyplot` 1189 to 1192 passing, 3 skipped. `make check` passes, 8 checks, 3452 passing. `ty` still 77 diagnostics, unchanged from base. The three earlier fixes re-verified intact: patch stroke `[5,5]` at 72 dpi and `[13,13]` at 200, annular sector fills at 0.5022, square-with-hole abstains.

Three tests added and one rewritten, failing against the previous head on the collapse itself:

```
FAILED test_patch_at_genomic_coordinates_keeps_every_vertex
FAILED test_tightly_spaced_but_distinct_vertices_survive
```

`test_full_disc_wedge_fills_despite_its_repeated_vertex` guards against the exact-equality remedy specifically, so it passes at head by construction. Mutation-tested rather than trusted: it fails against exact-only dedupe and against pre-dedupe head, both with `len(meshes) == 0`.

No changelog entry for either. Both were introduced inside this unreleased PR and never shipped, so relative to released `xy` a patch at 1e9 coordinates worked before and works now.

**Comment by michael-denyer** [comment]
Right, fixed in `2d66155`. Reproduced by calling `_without_repeats` directly:

```
                 before          after
finite           5 -> 5          5 -> 5
NaN              5 -> 5          5 -> 5
+inf             5 -> 1          5 -> 5
-inf and +inf    5 -> 1          5 -> 5
1e200            5 -> 5          5 -> 5
```

An infinite span makes `span * 1e-12` infinite, so every finite gap reads as a duplicate and the ring collapses to one vertex. NaN survived because comparisons against NaN are false, which is what the old comment described, so the comment was documenting half the behaviour. Early return on a non-finite span, as suggested, and the comment now says what the code does.

A genuine overflow does not reach this. `np.hypot` is overflow-safe, so a ring at 1e200 keeps a finite span. Infinity only arrives when the caller's geometry already contains it.

With the guard in place the negated comparison went back to a plain `gaps > tolerance`. The negation existed only to let NaN keep its vertex, and that job moved to the guard. It is now provably redundant rather than merely unused, since a non-finite coordinate propagates through `max` or `min` and forces a non-finite span, so a finite span implies finite gaps.

## On the test

The obvious input does not reach the guard. `Polygon([[0,0],[inf,0],[1,1],[0,1]])` never gets there, because matplotlib's `to_polygons` strips the non-finite vertex first:

```
inf raw rings: [[1.0, 1.0], [1.0, 1.0], [0.0, 1.0], [0.0, 0.0], [1.0, 1.0]]
```

`Rectangle((0, 0), inf, 1.0)` does reach it with all five vertices intact, so the test uses that and exercises the guard end to end through `add_patch` with a real matplotlib patch. Parametrised over `inf`, `-inf` and `nan`; both infinities fail against the previous head with the ring collapsed to a single vertex, and NaN passes either way, which is why it stays in the pair.

## Verification

`tests/pyplot` 1192 to 1195 passing, 3 skipped. `make check` passes, 8 checks, 3455 passing. `ty` still 77 diagnostics, unchanged from base across all seven commits. Earlier rounds re-verified intact: the 1e9 rectangle keeps area 50000.0, the full disc fills at 3.1414, the annulus abstains, patch stroke scales 5 px at 72 dpi to 13 px at 200.

No changelog entry. This one was introduced by the span tolerance in `e79b1da` earlier in this PR and never shipped.

**Comment by michael-denyer** [comment]
Right, fixed in `8701e42`. The comment described the code the guard replaced. Both `inf` and `nan` now take the same early return, so "opposite branches" was false and the comment was justifying the parametrisation with a reason that no longer existed.

Rewritten to what is actually true:

```python
# The tolerance is a fraction of the ring's span, and none of these three
# yield a usable one. They share a single guard, so what this pins is that
# its predicate catches all three: an isnan check would let the infinities
# through, an isinf check would let the NaN through.
```

Kept rather than deleted. The test name says the vertices survive but not why they would otherwise be lost, and "the tolerance is derived from the span" is not visible from the test body.

The new comment's claim is verified rather than asserted, by narrowing the guard in both directions:

```
guard narrowed to  if np.isnan(span):   -> FAILED [inf], FAILED [-inf]
guard narrowed to  if np.isinf(span):   -> FAILED [nan]
```

All three cases earn their place.

## Audit of the rest

Checked every comment, docstring and doc line added across the eight commits, since a guard that changes behaviour can strand prose written against the previous version.

One more found, a wording ambiguity rather than staleness. `_without_repeats` said "the test needs a tolerance", which reads as a unit test in a function whose behaviour is pinned by a test file, when it means the duplicate comparison. Changed to "the comparison".

Nothing else was stale. Each remaining comment states a measured number or a mechanism that still reproduces: the 1.000..4.000 unrotated span, 3.2509 for control-point ellipse area, 116 against a true 84 for the hole, the annular sector tracing a single ring, the swallowed rejection on the annulus, 1e-5 of 1e9 being 10,000 data units, the non-bit-exact repeat defeating exact-equality dedupe, and `np.isclose`'s 1e-8 floor swallowing a 1e-9 edge. `_shortest_relative_edge`'s stated bounds of 1e-16 and 1e-3 still bracket the measured 8.66e-17 and 6.96e-2.

One line re-checked specifically. `compat-changelog.md` names a full-circle `Wedge` with a `width` as a hole case. The withdrawn prediction was that such a wedge renders as a solid disc; the text claims it abstains, which still measures true at `rings=[33, 33] area=0.0000`. Left as written.

## Verification

No behaviour change and the numbers confirm it. `tests/pyplot` 1195 passed, 3 skipped, identical. `make check` passes, 8 checks, 3455 passing, identical. `ty` still 77, unchanged from base across all eight commits.

**Comment by Alek99** [comment]
Thanks for the contribution! There are couple tweaks im going to push but overall looks pretty good
